### PR TITLE
[TT-415] Add linting rule for optional response fields 

### PR DESCRIPTION
**Pull request by medric**
This PR introduces a new linting rule to ensure responses don't define optional fields within any object it may encapsulate. Nullable should be used instead. 

I've also done a bit of refactor to leverage composition an re-use small individual functions, as in extracting response types.

### CODE REVIEW

File: lib/src/linting/rules/no-optional-fields-within-responses.spec.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                    ]),
+                    optional: false
+                  }
+                ])
+              })
+            })
+          ]
+        })
+      ],
+      types: []
```
Comment (fwouts) [inline lib/src/linting/rules/no-optional-fields-within-responses.spec.ts]: I'd like to see some tests where a nested response type is a reference type as well (both valid and invalid case), to ensure that reference types are correctly handled.

File: lib/src/linting/rules/no-optional-fields-within-responses.spec.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  INT32,
+  NULL,
+  objectType,
+  STRING,
+  unionType
+} from "../../models/types";
+import { fakeLocatable } from "../../spec-helpers/fake-locatable";
+import { noOptionalFieldsWithinResponses } from "./no-optional-fields-within-responses";
+
+describe("rule: no optional fields within responses", () => {
```
Comment (fwouts) [inline lib/src/linting/rules/no-optional-fields-within-responses.spec.ts]: Both of the following can be interpreted as "optional":
```
{
  a?: string;
}

{
  a: string | null;
}
```

I prefer the term "omittable", which more clearly points to the former.

File: lib/src/linting/rules/no-optional-fields-within-responses.spec.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        })
+      ],
+      types: []
+    });
+    expect(errors).toEqual([
+      {
+        message:
+          "The object type `listUsers (response body for status 201)` defines an optional property. Use nullable instead."
+      },
+      {
```
Comment (fwouts) [inline lib/src/linting/rules/no-optional-fields-within-responses.spec.ts]: Why is it duplicated?

File: lib/src/utilities/extract-endpoint-types.ts [hunk 4]
```diff
@@ -2,15 +2,10 @@ import { compact } from "lodash";
 import { Locatable } from "../models/locatable";
 import { EndpointNode, TypeNode } from "../models/nodes";
 
-export function extractEndpointTypes(
+export function extractResponseTypes(
```
Comment (fwouts) [inline lib/src/utilities/extract-endpoint-types.ts:5]: It looks like both `extractRequestTypes` and `extractResponseTypes` return an array with at most one element. How about changing their return type to `TypeNode | null` instead?

File: lib/src/utilities/extract-nested-types.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
+import assertNever from "assert-never";
+import { flatten } from "lodash";
+import { TypeNode } from "../models/nodes";
+import { ObjectType, TypeKind, UnionType } from "../models/types";
+import { resolveType } from "../verifiers/utilities/type-resolver";
+
+export function extractNestedUnionTypes({
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]: It would be worth writing some JSDoc that explains that nested reference types are not included. This is a change of behaviour compared to the previous implementation.

File: lib/src/utilities/extract-nested-types.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    };
+  } catch (e) {
+    // Bubble up exception
+    throw e;
+  }
+}
+
+export function extractNestedObjectTypes(
+  typeNode: TypeNode,
+  typeStore: TypeNode[] = []
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]: I'd prefer to not use a default value here. It's very easy to omit the second argument by mistake.

**Review by fwouts — CHANGES_REQUESTED** [review]
Nice work.

File: lib/src/utilities/extract-endpoint-types.ts (hunk 4 above)
Comment (medric) [reply lib/src/utilities/extract-endpoint-types.ts:5]: `extractResponseTypes` can return an array with more than 1 element -> `endpoint.value.responses.map` and default response.

File: lib/src/linting/rules/no-omittable-fields-within-responses.spec.ts [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+              optional: false
+            }
+          ])
+        }
+      ]
+    });
+    expect(errors).toEqual([]);
+  });
+
+  test("rejects a response object when a field is optional instead of nullable", () => {
```
Comment (fwouts) [inline lib/src/linting/rules/no-omittable-fields-within-responses.spec.ts]: This still says "optional" instead of "omittable" :)

(other tests too)

File: lib/src/linting/rules/no-omittable-fields-within-responses.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    extractObjectTypes,
+    flatten
+  );
+
+  const types = flatten(contract.endpoints.map(extractTypes)) as Array<
+    TypeNode<ObjectType>
+  >;
+
+  return types.filter(hasOptionalProperties).map(typeNode => ({
+    message: `The object type \`${typeNode.name}\` defines an optional property. Use nullable instead.`
```
Comment (fwouts) [inline lib/src/linting/rules/no-omittable-fields-within-responses.ts]:
```suggestion
    message: `The object type \`${typeNode.name}\` defines an omittable property. Use nullable instead.`
```

File: lib/src/linting/rules/no-omittable-fields-within-responses.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  const extractObjectTypes = (t: TypeNode[]) =>
+    t.map((type: TypeNode) => extractNestedObjectTypes(type, contract.types));
+
+  const extractTypes = flow(
+    extractResponseTypes,
+    extractObjectTypes,
+    flatten
+  );
+
+  const types = flatten(contract.endpoints.map(extractTypes)) as Array<
```
Comment (fwouts) [inline lib/src/linting/rules/no-omittable-fields-within-responses.ts]: Why is a cast needed?

File: lib/src/utilities/extract-endpoint-types.ts (hunk 4 above)
Comment (fwouts) [reply lib/src/utilities/extract-endpoint-types.ts:5]: Good point. What about `extractRequestTypes`?

File: lib/src/utilities/extract-nested-types.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
+import assertNever from "assert-never";
+import { flatten } from "lodash";
+import { TypeNode } from "../models/nodes";
+import { ObjectType, TypeKind, UnionType } from "../models/types";
+import { resolveType } from "../verifiers/utilities/type-resolver";
+
+/**
+ * Recursively traverse the Types tree, find and group union types definitions into an flat array
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]:
```suggestion
 * Recursively traverse the Types tree, find and group union types definitions into a flat array.
 *
 * Note: Type references are not traversed.
```

File: lib/src/utilities/extract-nested-types.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+  try {
+    const unreferencedType = resolveType(typeNode.type, typeStore);
+
+    return {
+      name: typeNode.name,
+      type: unreferencedType
+    };
+  } catch (e) {
+    // Bubble up exception
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]: As far as I know:
```
try {
  [code]
} catch (e) {
  throw e;
}
```

is exactly equivalent to:
```
[code]
```

What's the reasoning for this?

File: lib/src/utilities/extract-nested-types.ts [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    throw e;
+  }
+}
+
+/**
+ * Recursively traverse the Types tree, find, resolve references, and group objects types definitions into an flat array
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]:
```suggestion
 * Recursively traverse the Types tree, find, resolve references, and group objects types definitions into a flat array.
```

File: lib/src/utilities/extract-nested-types.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+ */
+export function extractNestedObjectTypes(
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts:71]: This has a different behaviour from `extractNestedUnionTypes`: this one follows type refs, whereas the unions version doesn't. I think this is likely to trip up someone, no matter how well documented it is. Is there a good reason for the inconsistency?

**Review by fwouts — CHANGES_REQUESTED** [review]

File: lib/src/utilities/extract-nested-types.ts (hunk 13 above)
Comment (medric) [reply lib/src/utilities/extract-nested-types.ts:71]: Trying to find a way to avoid repeating logic, don't really now yet.

File: package.json [hunk 14]
```diff
@@ -24,6 +24,7 @@
     "fs-extra": "^8.1.0",
     "inquirer": "^6.5.0",
     "js-yaml": "^3.13.1",
+    "jsdoc": "3.6.2",
```
Comment (fwouts) [inline package.json]: Why was this added?

File: lib/src/utilities/extract-nested-types.ts (hunk 13 above)
Comment (fwouts) [reply lib/src/utilities/extract-nested-types.ts:71]: Let's chat in person.

File: lib/src/utilities/extract-endpoint-types.ts [hunk 15]
```diff
@@ -25,3 +20,25 @@ export function extractEndpointTypes(
     )
   ]);
 }
+
+export function extractRequestTypes(
```
Comment (fwouts) [inline lib/src/utilities/extract-endpoint-types.ts]:
```suggestion
export function extractRequestType(
```

File: lib/src/utilities/extract-endpoint-types.ts [hunk 16]
```diff
@@ -25,3 +20,25 @@ export function extractEndpointTypes(
...
+    null
+  );
+}
+
+export function extractEndpointTypes(
+  endpoint: Locatable<EndpointNode>
+): TypeNode[] {
+  return compact([
+    ...extractResponseTypes(endpoint),
+    extractRequestTypes(endpoint)
```
Comment (fwouts) [inline lib/src/utilities/extract-endpoint-types.ts]:
```suggestion
    extractRequestType(endpoint)
```

File: lib/src/utilities/extract-nested-types.ts [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+import { TypeNode } from "../models/nodes";
+import { ObjectType, TypeKind, UnionType } from "../models/types";
+import { resolveType } from "../verifiers/utilities/type-resolver";
+
+export function maybeResolveRef(
+  typeNode: TypeNode,
+  typeStore: TypeNode[]
+): TypeNode {
+  if (!typeStore.length) {
+    return typeNode;
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]: Shouldn't we crash when this occurs?

File: lib/src/utilities/extract-nested-types.ts [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    case TypeKind.STRING:
+    case TypeKind.FLOAT:
+    case TypeKind.DOUBLE:
+    case TypeKind.INT32:
+    case TypeKind.INT64:
+    case TypeKind.DATE:
+    case TypeKind.DATE_TIME:
+    case TypeKind.BOOLEAN_LITERAL:
+    case TypeKind.STRING_LITERAL:
+    case TypeKind.NUMBER_LITERAL:
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts:26]:
```suggestion
    case TypeKind.NUMBER_LITERAL:
      return [];
```

Comment (fwouts) [reply lib/src/utilities/extract-nested-types.ts:26]: (otherwise it will fall through!)

File: lib/src/utilities/extract-nested-types.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
+import assertNever from "assert-never";
+import { flatten } from "lodash";
+import { TypeNode } from "../models/nodes";
+import { ObjectType, TypeKind, UnionType } from "../models/types";
+import { resolveType } from "../verifiers/utilities/type-resolver";
+
+export function maybeResolveRef(
```
Comment (fwouts) [inline lib/src/utilities/extract-nested-types.ts]: Could we delete this function and use https://github.com/airtasker/spot/blob/master/lib/src/verifiers/utilities/type-resolver.ts#L41 instead?

File: package.json (hunk 14 above)
Comment (fwouts) [reply package.json]: Please address this comment too.

**Review by fwouts — CHANGES_REQUESTED** [review]

**Review by fwouts — CHANGES_REQUESTED** [review]
Please also add your linting rule to `rules.ts` :)

File: lib/src/utilities/extract-nested-types.ts (hunk 17 above)
Comment (medric) [reply lib/src/utilities/extract-nested-types.ts]: why ?

**Review by fwouts — APPROVED** [review]

**Review by lfportal — APPROVED** [review]
